modules/raft/src/main/java/org/apache/ignite/internal/raft/server/impl/JraftServerImpl.java

```diff
@@ -830,8 +785,12 @@ public void result(Serializable res) {
                     st = new Status(RaftError.ESTATEMACHINE, "Unknown state machine error.");
                 }
 
-                if (iter.done() != null) {
-                    iter.done().run(st);
+                // This is necessary so that IndexOutOfBoundsException is not thrown in a situation where the listener, when processing a
+                // command, catch any exception and does clo.result(throwable) and then throws the caught exception.
```

Comment:
Thanks for the explaining comment! Could you please expand it a bit: say that `clo.result(throwable)` actually advances the iterator, so next call to `done()` would have caused an out-of-bounds access.